Add bang patterns to IR

Daniel-Teut opened the pull request:
Extends the IR of the compiler to add bang patterns and change the generation of helper functions, so that these bang patterns are used instead of the environment. Additionally change the function translation to also use the bang patterns. Closes #23.

Daniel-Teut committed:
Add `isBang` flag to `VarPat` #23

Daniel-Teut committed:
Add `isBang` to `Alt` #23

Daniel-Teut committed:
Extend `pretty` instances for BangPatterns #23

Daniel-Teut committed:
Adjust test cases for extended AST #23

Daniel-Teut committed:
Extend parser for BangPatterns #23

Daniel-Teut committed:
Adjust pretty instance of `Alt` #23

The AST changed slightly.

Daniel-Teut committed:
Add information about strict arguments to `FuncEntry` #23

Daniel-Teut committed:
Rename `isBang` to `isStrict` #23

Daniel-Teut committed:
Update translation rules for function applications #23

Modify the genration of helper functions, so strict arguments are used. 
Further remove now redundant environment modifications.

Daniel-Teut committed:
Add test cases for the IR Parser #23

Daniel-Teut committed:
Format code with brittany #23

just95 commented on src/lib/FreeC/IR/Syntax/Expr.hs:
```diff
@@ -276,21 +276,30 @@ prettyExprPred' _ (Undefined _ _      ) = prettyString "undefined"
   , altRhs     :: Expr
+  , altIsStrict  :: Bool
   }
  deriving (Eq, Show)
```
After reading #23 again, I've got to admit that my wording was a bit ambiguous. With "bang patterns should also be supported in `case` expressions" I actually meant only variable patterns and not constructor patterns in `case` expressions. I originally did't consider constructor patterns since bang patterns only have an effect if they precede a variable pattern. Thus, I think that `altIsStrict` can be removed again. I'm sorry for the inconvenience. I'll try to be more clear in future issue descriptions.

just95 commented on src/lib/FreeC/Backend/Coq/Converter/Arg.hs:
```diff
@@ -84,7 +84,7 @@ convertArg (IR.VarPat srcSpan ident maybeArgType) = do
+convertPureArg (IR.VarPat srcSpan ident maybeArgType _) = do
   ident'        <- renameAndDefineVar srcSpan True ident maybeArgType
   maybeArgType' <- mapM convertType' maybeArgType
   generateArgBinder ident' maybeArgType'
```
One advantage of the bang pattern notation is that we would only need `convertArg`. If `varPatIsStrict` is `True`, `convertArg` should behave like `convertPureArg`.

just95 commented on src/lib/FreeC/Backend/Coq/Converter/Expr.hs:
```diff
@@ -117,7 +117,7 @@ convertExpr' (IR.Var srcSpan name _) typeArgs args = do
       -- Is this a recursive helper function?
       Just arity   <- inEnv $ lookupArity IR.ValueScope name
-      mDecArgIndex <- inEnv $ lookupDecArgIndex name
+      mDecArgIndex <- inEnv $ lookupFirstStrictArgIndex name
```
Even though there is at most one strict argument at the moment, you shouldn't make that assumption here since we may decide to use bang patterns for other purposes in the future. If a function is strict in two arguments, for example, both arguments must be bound before the function is invoked. Also add test cases for this behavior.

just95 commented on src/lib/FreeC/Backend/Coq/Converter/Expr.hs:
```diff
@@ -295,6 +295,6 @@ convertConPat (IR.ConPat srcSpan ident) varPats = do
 convertVarPat :: IR.VarPat -> Converter Coq.Pattern
-convertVarPat (IR.VarPat srcSpan ident maybeVarType) = do
+convertVarPat (IR.VarPat srcSpan ident maybeVarType _) = do
   ident' <- renameAndDefineVar srcSpan False ident maybeVarType
```
The flag of `renameAndDefineVar` that is set to `False` here indicates whether a value is in the `Free` monad or not. If the variable is annotated with a bang pattern, it should not be lifted. Thus, you have to pass `varPatIsStrict` to `renameAndDefineVar` such that the compiler knows that the variable needs to be wrapped with `pure` if it occurs in a position where a lifted value is expected. Also a bind needs to be inserted in `convertAlt` to implement the semantic of the bang pattern. Add test cases for bang patterns in `case` expressions.

just95 commented on src/lib/FreeC/Backend/Coq/Converter/FuncDecl/Rec/WithHelpers.hs:
```diff
@@ -153,16 +148,18 @@ transformRecFuncDecl (IR.FuncDecl srcSpan declIdent typeArgs args maybeRetType e
         (IR.VarPat NoSrcSpan . fromJust . IR.identFromQName)
         helperArgNames
         helperArgTypes
+        (map (== decArg) helperArgNames)
```
This could be a problem if arguments of the original IR function are marked as strict already.

just95 commented on src/lib/FreeC/IR/Similar.hs:
```diff
@@ -363,7 +363,7 @@ instance Similar IR.Alt where
 --   > ———————————  ——————————————————————————
 --   >  Γ ⊢ x ≈ y    Γ ⊢ (x :: τ) ≈ (y :: τ')
 instance Similar IR.VarPat where
-  similar' (IR.VarPat _ _ t1) (IR.VarPat _ _ t2) = similar' t1 t2
+  similar' (IR.VarPat _ _ t1 _) (IR.VarPat _ _ t2 _) = similar' t1 t2
```
Two variable patterns are not similar if only one of them has a bang pattern. The following two additional inference rules should hold.

```haskell
--   > —————————————  ————————————————————————————
--   >  Γ ⊢ !x ≈ !y    Γ ⊢ !(x :: τ) ≈ !(y :: τ')
```

Daniel-Teut committed:
Adjust `Similar` instance for `VarPat` #23

Two variable patterns are only similar if both have a bang pattern.

Daniel-Teut committed:
Adjust helper function generation #23

The strictness of the arguments of the original function is now 
preserved.

Daniel-Teut committed:
Add hlint suggestion #23

Daniel-Teut committed:
Change `convertArg` to use bang patterns #23

`convertArg` and `convertPureArg` can be the same function, since now 
the strictness annotation dictates, if the argument is lifted or not.

stu204767 commented on src/lib/FreeC/Backend/Coq/Converter/FuncDecl/Rec/WithHelpers.hs:
```diff
@@ -153,16 +149,23 @@ transformRecFuncDecl (IR.FuncDecl srcSpan declIdent typeArgs args maybeRetType e
         helperArgNames = Set.toList (usedVars `Set.intersection` boundVars)
 
     -- Determine the type of helper function's arguments and its return type.
+    -- Additionally the decreasing argument is marked as strict.
```
```suggestion
    -- Additionally, the decreasing argument is marked as strict.
```

stu204767 commented on src/lib/FreeC/IR/Similar.hs:
```diff
@@ -362,8 +362,15 @@ instance Similar IR.Alt where
 --   > ———————————  ——————————————————————————
 --   >  Γ ⊢ x ≈ y    Γ ⊢ (x :: τ) ≈ (y :: τ')
+--
+--   If one of the patterns has a bang pattern the other one needs one aswell.
```
```suggestion
--   If one of the patterns has a bang pattern, the other one needs one as well.
```

stu204767 commented on src/lib/FreeC/IR/Syntax/Expr.hs:
```diff
@@ -321,11 +330,13 @@ instance Pretty ConPat where
 --
---   The variable pattern can optionally have a type signature.
+--   The variable pattern can optionally have a type signature
+--   or can be annotated by a @!@.
```
```suggestion
--   or be annotated by a @!@.
```

stu204767 commented on src/test/FreeC/IR/SimilarTests.hs:
```diff
@@ -268,6 +268,10 @@ testSimilarFuncDecls = context "function declarations" $ do
     fx <- expectParseTestFuncDecl "f (x :: Foo) = x"
     fy <- expectParseTestFuncDecl "f y = y"
     fx `shouldNotBeSimilarTo` fy
+  it "functions with strict and non strict arguments are not similar" $ do
```
```suggestion
  it "functions with strict and non-strict arguments are not similar" $ do
```

stu204767 commented on src/test/FreeC/Frontend/IR/ParserTests.hs:
```diff
@@ -368,56 +368,107 @@ testIfExprParser = context "if expressions" $ do
+                      , IR.Alt NoSrcSpan barPat [] y False
                       ]
                       Nothing
+  it "accepts case expressions with one strict and one non strict alternative"
```
```suggestion
  it "accepts case expressions with one strict and one non-strict alternative"
```

stu204767 commented on src/test/FreeC/Frontend/IR/ParserTests.hs:
```diff
@@ -368,56 +368,107 @@ testIfExprParser = context "if expressions" $ do
+                      [IR.Alt NoSrcSpan fooPat [xPatStrict, yPatStrict] x False]
+                      Nothing
+  it
+      "accepts case expressions with strict variable patterns and strict alternative"
```
```suggestion
  it "accepts case expressions with strict variable patterns and a strict alternative"
```
(The line break might be added again by Brittany)

stu204767 commented on src/test/FreeC/Frontend/IR/ParserTests.hs:
```diff
@@ -368,56 +368,107 @@ testIfExprParser = context "if expressions" $ do
-                            [IR.Alt NoSrcSpan fooPat [xPat'] x]
+                            [IR.Alt NoSrcSpan fooPat [xPat'] x False]
                             Nothing
+  it "accepts case expressions with strict type annotated variable patterns"
```
```suggestion
  it "accepts case expressions with strict type-annotated variable patterns"
```

stu204767 commented on src/test/FreeC/Frontend/IR/ParserTests.hs:
```diff
@@ -508,22 +559,38 @@ testIntLiteralParser = context "integer literals" $ do
+                                [xPatStrict, yPatStrict]
+                                Nothing
+                                x
+  it "accepts function declarations with a type annotated argument" $ do
```
```suggestion
  it "accepts function declarations with a type-annotated argument" $ do
```

stu204767 commented on src/test/FreeC/Frontend/IR/ParserTests.hs:
```diff
@@ -508,22 +559,38 @@ testIntLiteralParser = context "integer literals" $ do
+                                x
+  it "accepts function declarations with a type annotated argument" $ do
     "f (x :: a) = x" `shouldParse` IR.FuncDecl NoSrcSpan f [] [xPat'] Nothing x
+  it "accepts function declarations with a strict type annotated argument" $ do
```
```suggestion
  it "accepts function declarations with a strict type-annotated argument" $ do
```

stu204767 reviewed (changes requested):
I made some minor suggestions (mostly missing hyphens in comments), but your code looks very good in total.

Daniel-Teut committed:
Remove `convertArgs` #23

Mapping over `convertArg` now has the same effect, since the explicit 
decreasing argument index is not needed anymore.

Daniel-Teut committed:
Remove `altIsStrict` from case alternatives #23

Daniel-Teut committed:
Convert bind generation for multiple strict arguments #23

Daniel-Teut committed:
Add test cases for multiple strict arguments #23

Daniel-Teut committed:
Remove redundant import

Daniel-Teut committed:
Apply suggestions from code review

Co-authored-by: stu204767 <63296935+stu204767@users.noreply.github.com>

Daniel-Teut committed:
Add bind generation to `convertAlt` #23

stu204767 committed:
Add tests for strict arguments in non-recursive functions #23

stu204767 committed:
Merge branch 'issue-23' into issue-23-tests

just95 commented on src/test/FreeC/Test/Environment.hs:
```diff
@@ -151,12 +152,14 @@ defineTestVar nameStr = do
+--   defined function is partial or not. The first argument controls the
+--   strictness of the function arguments.
+defineTestFunc' :: Bool -> [Bool] -> String -> Int -> String -> Converter String
+defineTestFunc' partial areStrict nameStr arity typeStr = do
```
The documentation does not match the implementation.

```suggestion
-- | Like 'defineTestFunc' but the first argument controls whether the
--   defined function is partial or not. The second argument controls the
--   strictness of the function arguments.
defineTestFunc' :: Bool -> [Bool] -> String -> Int -> String -> Converter String
defineTestFunc' partial areStrict nameStr arity typeStr = do
```

just95 commented on src/test/FreeC/Test/Environment.hs:
```diff
@@ -176,7 +180,13 @@ defineTestFunc' partial nameStr arity typeStr = do
+-- | Like 'defineTestFunc' but also allows to mark arguments as strict in the
+--   first argument.
+defineStrictTestFunc :: [Bool] -> String -> Int -> String -> Converter String
+defineStrictTestFunc = defineTestFunc' False
```
Since the list contains one element for every argument, the `arity` wouldn't have to be passed to this function explicitly.

```haskell
defineStrictTestFunc :: String -> [Bool] -> String -> Converter String
```

just95 commented on src/lib/FreeC/IR/Syntax/Expr.hs:
```diff
@@ -321,11 +321,13 @@ instance Pretty ConPat where
   { varPatSrcSpan :: SrcSpan
   , varPatIdent   :: String
   , varPatType    :: Maybe Type
+  , varPatIsStrict  :: Bool
```
```suggestion
data VarPat = VarPat
  { varPatSrcSpan   :: SrcSpan
  , varPatIdent     :: String
  , varPatType      :: Maybe Type
  , varPatIsStrict  :: Bool
```

just95 commented on src/lib/FreeC/IR/Similar.hs:
```diff
@@ -362,8 +362,15 @@ instance Similar IR.Alt where
 instance Similar IR.VarPat where
-  similar' (IR.VarPat _ _ t1) (IR.VarPat _ _ t2) = similar' t1 t2
+  similar' (IR.VarPat _ _ t1 s1) (IR.VarPat _ _ t2 s2) renaming =
+    s1 == s2 && similar' t1 t2 renaming
```
In the other `Similar` instances a point-free notation is used.

```suggestion
  similar' (IR.VarPat _ _ t1 s1) (IR.VarPat _ _ t2 s2) =
    const (s1 == s2) .&&. similar' t1 t2
```

just95 commented on src/lib/FreeC/Backend/Coq/Converter/Expr.hs:
```diff
@@ -115,26 +116,18 @@ convertExpr' (IR.Var srcSpan name _) typeArgs args = do
+      let typeArgNames = map (IR.UnQual . IR.Ident) typeArgIdents
+          subst = composeSubsts (zipWith singleSubst typeArgNames typeArgs)
+          decArgTypes = map (applySubst subst) argTypes
+      decArgTypes' <- mapM (mapM convertType') decArgTypes
```
The name `decArgTypes` doesn't really fit anymore.

stu204767 committed:
Add tests for strict patterns in case expressions #23

Daniel-Teut committed:
Merge branch 'master' into issue-23

stu204767 committed:
Add tests for strict arguments in recursive functions #23

stu204767 committed:
Merge branch 'issue-23' into issue-23-tests

Daniel-Teut committed:
Apply review suggestions #23

Remove arity from `defineStrictTestFunc`, since it is already implicitly 
passed in the length of the strict argument list.
Change mentions of `decArg` to `strictArg`.
Change `Similar` instance of `VarPat` to a point free style.

stu204767 commented on src/lib/FreeC/Environment.hs:
```diff
@@ -138,6 +137,20 @@ removeDecArg :: IR.QName -> Environment -> Environment
+-- | Changes the Coq identifier of the entry with the given name in the given
+--   scope to the given identifier.
+--
+--   If such an entry does not exists, the environment is not changed.
```
```suggestion
--   If such an entry does not exist, the environment is not changed.
```

stu204767 commented on src/lib/FreeC/Backend/Coq/Converter/Free.hs:
```diff
@@ -61,6 +61,40 @@ genericApply func effectArgs implicitArgs args
+--   The fresh variables and the not bound values are passed to the given
+--   function in their original order. The fresh variables are not visible
+--   outside this function. If a value is a variable, the name of that variable
+--   is used as the prefix of the corresponding fresh variable. Otherwise the
```
```suggestion
--   is used as the prefix of the corresponding fresh variable. Otherwise, the
```

Daniel-Teut committed:
Apply suggestions from code review

Co-authored-by: stu204767 <63296935+stu204767@users.noreply.github.com>

Daniel-Teut committed:
Merge branch 'master' into issue-23

just95 commented on src/lib/FreeC/Backend/Coq/Converter/Expr.hs:
```diff
@@ -115,26 +116,18 @@ convertExpr' (IR.Var srcSpan name _) typeArgs args = do
+      let typeArgNames = map (IR.UnQual . IR.Ident) typeArgIdents
+          subst = composeSubsts (zipWith singleSubst typeArgNames typeArgs)
+          strictArgTypes = map (applySubst subst) argTypes
+      strictArgTypes' <- mapM convertType' strictArgTypes
```
The name `strictArgTypes` isn't ideal either. Even though only the elements are used where `strictArgs` is set to `True`, the list itself does not just contain types of `strictArgs` but of all `args`. I'd suggest the following.

```suggestion
      argTypes' <- mapM (convertType' . applySubst subst) argTypes
```

just95 commented on src/test/FreeC/Test/Environment.hs:
```diff
@@ -176,7 +180,14 @@ defineTestFunc' partial nameStr arity typeStr = do
+
+-- | Like 'defineTestFunc' but also allows to mark arguments as strict in the
+--   first argument.
+defineStrictTestFunc :: [Bool] -> String -> String -> Converter String
+defineStrictTestFunc areStrict nameStr =
```
I would prefer the order of arguments to be swapped such that the function is consistent with how you usually write a function declaration in IR, i.e, `<name> <args> :: <type>`.

```suggestion
defineStrictTestFunc :: String -> [Bool] -> String -> Converter String
defineStrictTestFunc nameStr areStrict =
```

just95 commented on src/test/FreeC/Backend/Coq/Converter/FuncDecl/Rec/WithHelpersTests.hs:
```diff
@@ -501,6 +501,83 @@ testConvertRecFuncDeclWithHelpers = context "with helper functions" $ do
+                                       2
+                                       "forall a. List a -> List a -> List a"
+        shouldConvertWithHelpersTo
+            [ "interleave @a (xs :: List a) !(ys :: List a) :: List a ="
```
I think it would be also interesting to have an example where the bang pattern precedes the decreasing argument. Such an example could reveal a problem in the generation of the `struct` annotation.

Daniel-Teut committed:
Apply review suggestions

Co-Authored-By: Justin Andresen <jan.dresen.95@gmail.com>

Daniel-Teut committed:
Change `defineTestFunc` to its strict version where appropriate #23

Daniel-Teut committed:
Add test case for strict argument preceding decreasing argument #23

Daniel-Teut committed:
Fix bug in `struct` generation #23

If a strict argument preceded the decreasing argument, this strict 
argument was used for the `struct` generation. Now use the correct 
argument.
